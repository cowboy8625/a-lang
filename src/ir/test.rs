// use crate::semantic_analysis::{SemanticAnalysisVisitor, SymbolTableBuilder};
macro_rules! snapshot {
    ($name:tt, $path:tt) => {
        #[test]
        fn $name() {
            use super::*;
            use $crate::lexer::lex;
            use $crate::parse::parse;
            let contents = include_str!($path);
            let tokens = lex(contents).unwrap();
            let ast = parse(tokens).unwrap();
            // let mut symbol_table_builder = SymbolTableBuilder::default();
            // symbol_table_builder.visit(&ast);
            // let symbol_table = symbol_table_builder.build();
            let ir_code = code_gen(ast).unwrap();
            let result = ir_code
                .iter()
                .map(ToString::to_string)
                // .map(|i| format!("{i:#?}\n"))
                .collect::<String>();
            let mut settings = insta::Settings::clone_current();
            settings.set_snapshot_path("testdata/output/");
            settings.bind(|| {
                insta::assert_snapshot!(result);
            });
        }
    };
}

snapshot!(binary, "../../snapshots/binary.a");
snapshot!(ifelse, "../../snapshots/ifelse.a");
snapshot!(max, "../../snapshots/max.a");
