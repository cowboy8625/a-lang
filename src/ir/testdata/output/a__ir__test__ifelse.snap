---
source: src/ir/test.rs
expression: result
---
DefFunc(
    DefFunc {
        name: "main",
        ret: I64,
        params: [],
        body: [
            Enter(
                Enter,
            ),
            LoadImm(
                LoadImm {
                    des: Reg(
                        0,
                    ),
                    imm: Imm(
                        1,
                    ),
                },
            ),
            LoadImm(
                LoadImm {
                    des: Reg(
                        1,
                    ),
                    imm: Imm(
                        3,
                    ),
                },
            ),
            Grt(
                Grt {
                    des: Reg(
                        2,
                    ),
                    lhs: Reg(
                        0,
                    ),
                    rhs: Reg(
                        1,
                    ),
                },
            ),
            Conditional(
                Conditional {
                    label: Label(
                        ".L0",
                    ),
                    reg: Reg(
                        2,
                    ),
                },
            ),
            LoadImm(
                LoadImm {
                    des: Reg(
                        3,
                    ),
                    imm: Imm(
                        1,
                    ),
                },
            ),
            Return(
                Return(
                    Reg(
                        3,
                    ),
                ),
            ),
            Jump(
                Jump(
                    Label(
                        ".exit",
                    ),
                ),
            ),
            DefLabel(
                DefLabel(
                    Label(
                        ".L0",
                    ),
                ),
            ),
            LoadImm(
                LoadImm {
                    des: Reg(
                        4,
                    ),
                    imm: Imm(
                        100,
                    ),
                },
            ),
            Return(
                Return(
                    Reg(
                        4,
                    ),
                ),
            ),
            Jump(
                Jump(
                    Label(
                        ".exit",
                    ),
                ),
            ),
            DefLabel(
                DefLabel(
                    Label(
                        ".exit",
                    ),
                ),
            ),
            Leave(
                Leave,
            ),
        ],
    },
)

